File: sdks/js/src/credentials.js
/**
 * Credential storage helpers — Node stdlib only.
 *
 * Used by both the high-level client (`client.ts`) and the stdlib-only
 * MCP CLI bridge (`cli.js`). Plain JS so the bridge runs without any
 * TypeScript tooling and consumers don't pay a compile cost for it.
 *
 * Storage location:
 *   - macOS/Linux: $XDG_CONFIG_HOME/ailang-parse/credentials.json
 *                  (default ~/.config/ailang-parse/credentials.json)
 *   - Windows:     %APPDATA%\ailang-parse\credentials.json
 *
 * File format:
 *   {
 *     "api_key":  "dp_...",
 *     "base_url": "https://docparse.ailang.sunholo.com",
 *     "key_id":   "...",
 *     "tier":     "free",
 *     "label":    "..."
 *   }
 */

import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { homedir, platform } from "node:os";
import { join } from "node:path";

export const DEFAULT_BASE_URL = "https://docparse.ailang.sunholo.com";
const CONFIG_DIR_NAME = "ailang-parse";
const CREDENTIALS_FILE = "credentials.json";

/** Return the platform-appropriate config directory, or null in browsers. */
export function configDir() {
  try {
    if (platform() === "win32") {
      const base = process.env.APPDATA || join(homedir(), "AppData", "Roaming");
      return join(base, CONFIG_DIR_NAME);
    }
    const base = process.env.XDG_CONFIG_HOME || join(homedir(), ".config");
    return join(base, CONFIG_DIR_NAME);
  } catch {
    return null; // browser — no filesystem
  }
}

/** Return the absolute credentials file path, or null in browsers. */
export function credentialsPath() {
  const dir = configDir();
  return dir ? join(dir, CREDENTIALS_FILE) : null;
}

/**
 * Load saved credentials matching `baseUrl`. Returns null if absent or
 * the saved entry is for a different base_url.
 */
export function loadSavedKey(baseUrl = DEFAULT_BASE_URL) {
  try {
    const path = credentialsPath();
    if (!path || !existsSync(path)) return null;
    const data = JSON.parse(readFileSync(path, "utf-8"));
    if (data?.api_key?.startsWith("dp_") && (data.base_url || DEFAULT_BASE_URL) === baseUrl) {
      return data;
    }
  } catch {}
  return null;
}

/**
 * Resolve any saved API key from env var or credentials file.
 *
 * Unlike `loadSavedKey`, this does NOT filter by base_url — it returns
 * whatever key is on disk. Used by the MCP CLI bridge, which forwards
 * to whatever endpoint the user configured via AILANG_PARSE_MCP_URL
 * and just needs *a* key to inject.
 */
export function resolveApiKey() {
  if (process.env.DOCPARSE_API_KEY) return process.env.DOCPARSE_API_KEY;
  try {
    const path = credentialsPath();
    if (!path || !existsSync(path)) return null;
    const data = JSON.parse(readFileSync(path, "utf-8"));
    if (typeof data?.api_key === "string" && data.api_key.startsWith("dp_")) {
      return data.api_key;
    }
  } catch {}
  return null;
}

/** Persist credentials to disk with restrictive permissions (0600 file, 0700 dir). */
export function saveKey(apiKey, baseUrl = DEFAULT_BASE_URL, keyId = "", tier = "free", label = "") {
  try {
    const dir = configDir();
    if (!dir) return;
    mkdirSync(dir, { recursive: true, mode: 0o700 });
    const path = join(dir, CREDENTIALS_FILE);
    const payload = JSON.stringify(
      { api_key: apiKey, base_url: baseUrl, key_id: keyId, tier, label },
      null,
      2,
    ) + "\n";
    writeFileSync(path, payload, { mode: 0o600 });
  } catch {
    // ignore — browser or permission error
  }
}

File: sdks/r/R/credentials.R
#' Credential storage helpers
#'
#' Mirrors the cross-SDK credential format used by the Python, JavaScript and
#' Go AILANG Parse clients so a key saved in any one of them is usable from
#' any other. The on-disk file is JSON with these fields:
#' \code{api_key}, \code{base_url}, \code{key_id}, \code{tier}, \code{label}.
#'
#' Storage location:
#' \itemize{
#'   \item Linux/macOS: \code{$XDG_CONFIG_HOME/ailang-parse/credentials.json}
#'         (default \code{~/.config/ailang-parse/credentials.json}).
#'   \item Windows: \code{\%APPDATA\%/ailang-parse/credentials.json}.
#' }
#'
#' On Unix the file is written with mode \code{0600} and the directory with
#' \code{0700}.
#'
#' @name credentials
NULL

#' @describeIn credentials Default hosted API endpoint.
#' @return \code{default_base_url()} returns a single string.
#' @examples
#' default_base_url()
#' @export
default_base_url <- function() {
  "https://docparse.ailang.sunholo.com"
}

.config_dir_name <- "ailang-parse"
.credentials_file <- "credentials.json"

.config_dir <- function() {
  if (.Platform$OS.type == "windows") {
    base <- Sys.getenv("APPDATA")
    if (!nzchar(base)) {
      base <- file.path(Sys.getenv("USERPROFILE"), "AppData", "Roaming")
    }
  } else {
    base <- Sys.getenv("XDG_CONFIG_HOME")
    if (!nzchar(base)) {
      base <- file.path(path.expand("~"), ".config")
    }
  }
  file.path(base, .config_dir_name)
}

#' @describeIn credentials Absolute path to the credentials file (whether or
#'   not it exists).
#' @examples
#' credentials_path()
#' @export
credentials_path <- function() {
  file.path(.config_dir(), .credentials_file)
}

.read_credentials_file <- function() {
  path <- credentials_path()
  if (!file.exists(path)) return(NULL)
  data <- tryCatch(
    jsonlite::fromJSON(path, simplifyVector = TRUE),
    error = function(e) NULL
  )
  if (!is.list(data)) return(NULL)
  key <- data$api_key
  if (!is.character(key) || length(key) != 1L || !startsWith(key, "dp_")) {
    return(NULL)
  }
  data
}

#' @describeIn credentials Load saved credentials matching \code{base_url}, or
#'   \code{NULL} if no compatible record exists.
#' @param base_url Base URL the saved key must match.
#' @examples
#' \dontrun{
#' load_saved_key()
#' }
#' @export
load_saved_key <- function(base_url = default_base_url()) {
  data <- .read_credentials_file()
  if (is.null(data)) return(NULL)
  saved_url <- if (is.null(data$base_url) || !nzchar(data$base_url)) {
    default_base_url()
  } else {
    data$base_url
  }
  if (!identical(saved_url, base_url)) return(NULL)
  data
}

#' @describeIn credentials Persist credentials to disk with restrictive
#'   permissions. Returns the path written, invisibly.
#' @param api_key API key string (must start with \code{dp_}).
#' @param key_id Opaque key identifier returned by the device-auth flow.
#' @param tier Plan tier (\code{"free"}, \code{"pro"}, ...).
#' @param label Human-readable label for the key.
#' @examples
#' \dontrun{
#' save_key("dp_your_key_here", label = "my-laptop")
#' }
#' @export
save_key <- function(api_key,
                     base_url = default_base_url(),
                     key_id = "",
                     tier = "free",
                     label = "") {
  stopifnot(is.character(api_key), length(api_key) == 1L, nzchar(api_key))
  d <- .config_dir()
  if (!dir.exists(d)) {
    dir.create(d, recursive = TRUE, showWarnings = FALSE)
  }
  if (.Platform$OS.type != "windows") {
    Sys.chmod(d, mode = "0700")
  }
  payload <- list(
    api_key  = jsonlite::unbox(api_key),
    base_url = jsonlite::unbox(base_url),
    key_id   = jsonlite::unbox(key_id),
    tier     = jsonlite::unbox(tier),
    label    = jsonlite::unbox(label)
  )
  path <- file.path(d, .credentials_file)
  json <- jsonlite::toJSON(payload, pretty = TRUE)
  writeLines(json, path)
  if (.Platform$OS.type != "windows") {
    Sys.chmod(path, mode = "0600")
  }
  invisible(path)
}

#' @describeIn credentials Resolve any saved API key — checks
#'   \code{DOCPARSE_API_KEY} first, then the credentials file (without
#'   filtering by base URL). Used by the MCP bridge.
#' @examples
#' resolve_api_key()
#' @export
resolve_api_key <- function() {
  env_key <- Sys.getenv("DOCPARSE_API_KEY", unset = "")
  if (nzchar(env_key)) return(env_key)
  data <- .read_credentials_file()
  if (is.null(data)) return(NULL)
  data$api_key
}
